spec.html

```diff
@@ -648,15 +543,134 @@ <h1>[[Construct]] ( _argumentsList_, _newTarget_)</h1>
+
+<emu-note type=editor>
+  <p>Although this specification uses logic of a [[PrivateBrand]] List which is held per-instance, the use of this list obeys certain invariants:</p>
+
+  <ul>
+    <li>Nothing is removed from the list; things are only added to the end.</li>
+    <li>There is no way to distinguish different copies of the list.</li>
+    <li>The list is mutable, but there are never any references to it outside of the particular object which holds it, so additions can be thought of as replacing the internal slot with another list which contains the new element.</li>
+  </ul>
+
+  <p>There are multiple possible implementation strategies which take advantage of these invariants for better runtime performance and lower space overhead, e.g., the following successive optimizations:</p>
+
+  <ul>
+    <li>The set of lists can be represented as a tree, where nodes represent a brand list, and the path to the root is the set of brands included. This tree will only have nodes added to it over time, and never needs to be mutated in any other way. In typical code, nodes only need to be allocated and added the first time a class is instantiated.</li>
+    <li>If deep inheritance hierarchies with private methods are common in practice, each node can have an array representing the set of brands, which can be traversed with better cache locality.</li>
+    <li>If a brand comes from a base class, only the first element of the array needs to be examined, rather than traversing up the tree.</li>
+    <li>If a brand comes from a subclass, the search can speculatively first check the entry at the depth in the class hierarchy (of classes with private methods) where the brand was at the time the class was created, and fall back to checking the whole array only on failure, along the lines of the classic Java O(1) instanceof check algorithm.</li>
+  </ul>
+</emu-note>
+
+<emu-clause id="sec-privatebrandcontains" aoid="PrivateBrandContains">
+  <h1>PrivateBrandCheck(_O_, _P_)</h1>
   <emu-alg>
-    1. Assert: _P_ is a Private Name value.
-    1. Assert: _desc_ is a Property Descriptor.
-    1. If _O_ is not an object, throw a *TypeError* exception.
-    1. Let _entry_ be PrivateFieldFind(_P_, _O_).
-    1. If _entry_ is not ~empty~, throw a *TypeError* exception.
-    1. Append { [[PrivateName]]: P, [[PrivateFieldDescriptor]]: _desc_ } to _O_.[[PrivateFieldDescriptors]].
+    1. Let _descriptor_ be _P_'s associated [[Descriptor]].
+    1. If _O_.[[PrivateBrands]] does not contain an entry _e_ such that SameValue(_e_, _descriptor_.[[Brand]]) is *true*,
+      1. Throw a *TypeError* exception.
+  </emu-alg>
+</emu-clause>
+
+<emu-clause id="sec-privatebrandadd" aoid="PrivateBrandAdd">
+  <h1>PrivateBrandAdd(_O_, _brand_)</h1>
+  <emu-alg>
+    1. If _O_.[[PrivateBrands]] contains an entry _e_ such that SameValue(_e_, _brand_) is *true*,
+      1. Throw a *TypeError* exception.
+    1. Append _P_ to _O_.[[PrivateBrands]].
```

Comment:
What is P?

Reply:
Oh, that should say _brand_, thanks, will fix.